org/mozilla/jss/provider/java/security/RSAPSSAlgorithmParameters.java

```diff
@@ -0,0 +1,244 @@
+    public final static BigInt          defaultSaltLen = new BigInt(20);
+    public final static BigInt          defaultTrailerField = new BigInt(1);
+
+    private PSSParameterSpec spec = PSSParameterSpec.DEFAULT;
+    private AlgorithmId hashAlg = defaultHashAlg;
+    private AlgorithmId maskGenFunc = defaultMaskGenFunc;
+    private BigInt saltLen = defaultSaltLen;
+    private BigInt trailerField = defaultTrailerField;
+
+    public RSAPSSAlgorithmParameters() {}
+
+    @Override
+    protected void engineInit(AlgorithmParameterSpec paramSpec)
+            throws InvalidParameterSpecException {
+        spec = (PSSParameterSpec) paramSpec;
+        populateFromSpec();
+    }
+
+    @Override
+    protected <T extends AlgorithmParameterSpec> T engineGetParameterSpec(Class<T> paramSpec)
+            throws InvalidParameterSpecException  {
+        if (paramSpec.isAssignableFrom(PSSParameterSpec.class)) {
+            return paramSpec.cast(spec);
+        }
+
+        throw new InvalidParameterSpecException("Unknown parameter spec passed to PSS parameters object: " + paramSpec.getName());
+    }
+
+    @Override
+    protected void engineInit(byte[] params) throws IOException {
+        decode(new DerInputStream(params), params);
+    }
+
+    @Override
+    protected void engineInit(byte[] params, String format) throws IOException {
+        // Assume Der for now.
+        Assert.notReached("engineInit(byte[],String) not supported");
+        throw new IOException("engineInit(byte[],String) not supported");
+    }
+
+    @Override
+    protected byte[] engineGetEncoded() throws IOException {
+        DerOutputStream out = new DerOutputStream();
+        encode(out);
+        return out.toByteArray();
+
+    }
+
+    @Override
+    protected byte[] engineGetEncoded(String format) throws IOException {
+        //Assume Der for now.
+        Assert.notReached("engineGetEncoded(String format)) not supported");
+        throw new IOException("engineGetEncoded(String format)) not supported");
+    }
+
+    @Override
+    protected String engineToString() {
+        String str = new String("Mozilla-JSS PSSAlgorithmParameters " +  getClass().getName() + " HashAlg: " + spec.getDigestAlgorithm() + " MaskGenAlg: " + spec.getMGFAlgorithm() );
+        return str;
+    }
+
+    private void decode(DerInputStream in , byte[] encoded) throws IOException {
+        if (in == null) {
+            throw new IOException("Invalid input: got null DerInputStream");
+        }
+
+        // Sequence has 3 members, trailer field ignored
+        DerValue seq[] = in.getSequence(3);
+        if (seq.length != 3) {
+            throw new IOException("Invalid data! Expected a sequence of at least 3 members.");
+        }
+
+        if (seq[0].isContextSpecific((byte)0)) {
+            seq[0] = seq[0].data.getDerValue();
+        } else {
+             throw new IOException("Invalid encoded data!");
```

Comment:
@jmagne While this is still fresh, do you have better descriptions here? I'm not sure I've fully grasped this DER decoding business here and its relevance.

This applies to lines 148 also.

Reply:
OK this context stuff has to do with the sequence of this thing, here is the blob again:

RSASSA-PSS-params ::= SEQUENCE {
   hashAlgorithm      [0] OAEP-PSSDigestAlgorithms  DEFAULT sha1,
   maskGenAlgorithm   [1] PKCS1MGFAlgorithms  DEFAULT mgf1SHA1,
   saltLength         [2] INTEGER  DEFAULT 20,
   trailerField       [3] INTEGER  DEFAULT 1


See, the 0,1,2, and 3 ?? This stuff goes into the encoding and is read out when decoding.. It took me many failures of constructing a legal blob before I looked into the context tags like here...